Tests for particle diffusion

jlubo opened the pull request:
Python tests are added to ensure the working of the diffusion of arbitrary particles in setups with 1-3 segments (cf. issues https://github.com/arbor-sim/arbor/issues/2084 and https://github.com/arbor-sim/arbor/issues/2145). 

Besides showing that the diffusion seems to work if segments have the same radius, the tests that are added here also point to the fact that the diffusion across segments of different radius is still erroneous. Therefore, I left the tests for different radii commented until https://github.com/arbor-sim/arbor/issues/2145 has been resolved.

thorstenhater commented:
Hey @jlubo,

thanks for adding this. Could you run `black .` and push the results? That'll make the Linter
check pass.

thorstenhater commented on python/test/unit/test_diffusion.py:
```diff
@@ -0,0 +1,315 @@
+            dec.discretization(
+                A.cv_policy(f"(fixed-per-branch {num_segs*num_cvs_per_seg} (branch 0))")
+            )
+        elif num_segs == 3:
```
Does this result in different numbers of CVs? It seems like the proposed
morphology produces either one linear or  three branches in Y-shape. In both cases you're setting `num_cvs_per_seg` on all branches. Thus, this should be equivalent to just setting `fixed-per-branch`?!

jlubo replied:
Thanks for pointing that out. Indeed the branch specification is unnecessary here, I've removed it from the code. But the distinction between `num_segs < 3` and `num_segs == 3` still is necessary because for less than three segments, there is only one branch.

thorstenhater replied:
True. Question though: Do you want a fixed number/branch or really a fixed $\Delta x$?

jlubo replied:
Yeah, I wanted to control the number of CVs, however, I guess here it shouldn't matter which quantity is kept fixed.

thorstenhater replied:
That seemed a bit unusual and so I am asking.

thorstenhater reviewed (changes requested):
Hey, 
made some comments for improvements, rest looks great, thanks.

jlubo commented on .gitignore:
```diff
@@ -105,3 +105,6 @@ _skbuild
 results
+
+# environment setting script
+set_arbor_env*
```
It's a file that I usually use to set the Arbor environment, I guess it can't be considered common :D

jlubo commented on python/test/unit/test_diffusion.py:
```diff
@@ -0,0 +1,315 @@
+      from concentrations by integrating over the volume of the morphology. The total amount
+      of particles should be conserved unless there is deliberate injection or removal of
+      particles.
+"""
```
:)

jlubo commented:
> Hey, made some comments for improvements, rest looks great, thanks.

Thanks @thorstenhater! I've revised the code according to the comments.

thorstenhater commented on ext/fmt:
I'd recommend `git fetch --all && git merge origin/master && git submodule sync` ... changing submodules in a feature PR is likely not what you want.

jlubo replied:
Thanks. I must have missed that they were in.

thorstenhater commented on ext/pugixml:
Same as above.

thorstenhater reviewed (approved):
Thank you very much!

thorstenhater commented:
cscs-ci run daint-gpu

thorstenhater commented:
Uh-oh
```
subprocess.CalledProcessError: Command 'make' returned non-zero exit status 2.
 Error:
/tmp/tmpqgxm6fo0/build/generated/diffusion/neuron_with_diffusion_gpu.cu(59): warning #177-D: function "arb::diffusion_catalogue::<unnamed>::multiply" was declared but never referenced
/tmp/tmpqgxm6fo0/build/generated/diffusion/synapse_with_diffusion_gpu.cu(74): error: identifier "n_" is undefined
/tmp/tmpqgxm6fo0/build/generated/diffusion/synapse_with_diffusion_gpu.cu(55): warning #177-D: function "arb::diffusion_catalogue::<unnamed>::multiply" was declared but never referenced
1 error detected in the compilation of "/tmp/tmpqgxm6fo0/build/generated/diffusion/synapse_with_diffusion_gpu.cu".
make[2]: *** [CMakeFiles/diffusion-catalogue.dir/build.make:183: CMakeFiles/diffusion-catalogue.dir/generated/diffusion/synapse_with_diffusion_gpu.cu.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/diffusion-catalogue.dir/all] Error 2
```

jlubo commented:
The different-radii and different-length tests are now uncommented. They should succeed if they are run with this new fix in the diffusion solver: https://github.com/thorstenhater/arbor/commit/7577092f89e9b54d8fde26ac6963a7c9965e43cc

thorstenhater commented:
Hey @jlubo,
 
could you leave them disabled, I'll add them to the to-be-made PR for the fixes.

jlubo commented:
Done :)

thorstenhater commented:
I'll fix the GPU problem in my follow-up #2226

thorstenhater commented on python/test/unit/test_diffusion.py:
```diff
@@ -351,31 +350,59 @@ def simulate_and_test_diffusion(
 
+        # additional tests for the case that there is only one segment of fixed radius
+        if num_segs == 1:
+            self.assertAlmostEqual(
```
In the varying radius case: Can't we estimate the equilibrium concentration as:
$N = \sum\limits_{i\in C} V_i Xi(t=0)$ where $C$ is the set of CV indices
and 
$X_i(t\rightarrow\infty) = \frac{N}{V_i}$
with 
$X_i(t\rightarrow\infty) = X_j(t\rightarrow\infty)\quad\forall i,j\in C$?

jlubo replied:
Yes! I've addressed this now, also see the comment below.

thorstenhater reviewed (changes requested):
Nice! See also downstream #2226. Two questions:
1. It seems like lots of special casing for the different setups (num_sgs=1,2,3). Do you think it would be cleaner to split these into different tests altogether?
2. Question also inline: Can't we always compute the equilibrium concentration?

jlubo commented:
>     1. It seems like lots of special casing for the different setups (num_sgs=1,2,3). Do you think it would be cleaner to split these into different tests altogether?

I agree, I've now split the function `get_morph_and_decor()` into into individual functions for 1, 2, and 3 segments.

>     2. Question also inline: Can't we always compute the equilibrium concentration?

Yes we can! I've introduced a more general computation of CV volumes making this possible now.

boeschf commented:
cscs-ci run daint-gpu

thorstenhater commented:
Hey @jlubo,

could I close this? It's subsumed in #2226.

jlubo commented:
Yes, sure. I'll check #2226 again in the next days.
